lib/puppet/util.rb

```diff
@@ -25,20 +25,65 @@ module Util
 
   # Run some code with a specific environment.  Resets the environment back to
   # what it was at the end of the code.
-  def self.withenv(hash)
-    saved = ENV.to_hash
-    hash.each do |name, val|
-      ENV[name.to_s] = val
-    end
-
+  # Windows can store unicode chars in the environment, so need to use API calls
+  #   to get and set the environment instead of ENV due to encoding problems.
+  def self.withenv(hash, mode=:posix)
+    saved = get_environment(mode)
+    set_environment(hash, mode)
     yield
   ensure
-    ENV.clear
-    saved.each do |name, val|
-      ENV[name] = val
+    restore_environment(saved, mode) if saved
+  end
+
+  def self.get_environment(mode=:posix)
+    case mode
+      when :posix
+        ENV.to_hash
+      when :windows
+        require 'puppet/util/windows/process'
+        Puppet::Util::Windows::Process.get_environment_strings
+      else
+        raise "Unable to retrieve the environment for mode #{mode}"
     end
   end
 
+  def self.set_environment(hash, mode=:posix)
+    case mode
+      when :posix
+        hash.each do |name, val|
+          ENV[name.to_s] = val
+        end
+      when :windows
+        require 'puppet/util/windows/process'
+        hash.each do |name, val|
+          Puppet::Util::Windows::Process.set_environment_variable(name,val)
+        end
+      else
+        raise "Unable to set the environment for mode #{mode}"
+    end
+  end
+
+  def self.restore_environment(saved, mode=:posix)
```

Comment:
So I generally like the refactor - makes it much clearer now IMHO.  A couple of minor nits though:

* `set_environment` should probably be renamed to `extend_environment` or similar (yeah, I realized I suggested the original name - but really weren't not setting anything, we're in reality appending, and method name should probably reflect that)
* `restore_environment` is really really close to the current impl of `set_environment` ... but should probably be named `set_environment`, since the goal is that the given hash is becoming the new environment.  That said, I think we can probably reuse the prior method like:

```ruby
def self.set_environment(hash, mode = :posix)
  case mode
  when :posix
    ENV.clear
    extend_environment(hash, mode)
  when :windows
    current = Puppet::Util::Windows::Process.get_environment_strings
    (current.keys - hash.keys).each { |extra_key| hash[extra_key] = nil }
    # extraneous env vars are set to nil, as this indicates they should be unset for Windows
    extend_environment(hash, :windows)
  else
    raise "Unable to set environment for mode #{mode}"
  end
end
```